Support template, jq flags with standard format flag

heaths opened the pull request:
Defines a new `cmdutil` package function that adds support for `--format` along with `--template` and `--jq`. This can be used instead of the `--json` flags which requires a field list and will filter the GraphQL query when filtering is unnecessary.

heaths committed:
Use standard JSON flags for project command

Deprecates the --format flag and adds the standard JSON flags to properly filter, template, and write JSON.

heaths commented:
@samcoe this is part of the work to standardize the `project` command. Before I go too far down this path for all the `project` subcommands, any concerns? The main issue is that the fields passed to `--json` are effectively ignored currently because of how queries are constructed. If everything else looks fine (and I'll keep going for all the subcommands that implement `--format json`), I can refactor the queries to support filtering in this PR or another.

There's two ideas, and I'm certainly open to alternatives of feedback:

1. Refactor the queries as string queries like all the other commands. Pretty much, get rid of all the `graphql`-tagged structs used as input. It's a big change, but probably better long term as we don't need separate request and response structs (which, that alone could probably be unified separately).
2. The number of fields requested isn't huge. Instead of modifying the queries, I could simply filter the fields as I'm writing them out with either a backward compatible tweak to the exporter, or something local to the `project` utility functions.

samcoe commented:
@heaths The main concerns here are that the `--format` flag is meant to be semantically different than the `--json` flag. The `--json` requires arguments for which fields to fetch, while `--format` is just meant to set the output. We designed it this way with the hopes of adding a `csv` format option for the `--format` flag as we anticipated that being a desired export format for projects so they can be imported in other tools. 

I think the best approach here is keep the `--format` flag as is but add support for the `--template` and `--jq` flags. Obviously this is still a slight variation from other commands but a good compromise. This will greatly reduce the amount of work needed to rewrite the queries to support filtering and keeps the command open for a `csv` output option in the future. What do you think?

heaths committed:
Add format flags cmdutil

Resolves PR feedback

heaths commented:
@samcoe, what do you think of this? Figured I'd make it general if `--format` is going to be a new thing in other cases. Also added color, though a bit redundant for the project state but it otherwise retains the current table info. If good (or minor suggestions), I'll give the same `AddFormatFlags` treatment to all the `project` subcommands and add a bit more color as well.

That said, if you want to retain the table header for TTY - which is inconsistent, but has long since shipped - what about making it a lighter gray, like dates? Personally, though I'd love to hear other feedback, I find the table headers distracting and harder to separate the table data at a glance. Making them gray - like they are less important but still provide some value - might strike a good UX balance. Thoughts? It could be implemented as a `WithOptions`-style option so only `project` subcommands get it for TTY.

E.g.,

<img width="269" alt="image" src="https://github.com/cli/cli/assets/1532486/c4363e9b-9981-43ce-bd5a-1dfaad6a04c4">

Even easier and consistent, I could do it in a way - capture `iostreams.ColorScheme()` on `tableprinter.New` - that works for all commands printing a header (which, now that I find them all, seems like it's the "new thing"):

<img width="397" alt="image" src="https://github.com/cli/cli/assets/1532486/be77890c-e128-4237-a8e4-177ea25240c6">

I could tweak the color a bit, too, so it doesn't exactly match the typical date fields.

Or maybe I should just do the colorizing in another PR.

samcoe commented on pkg/cmd/project/list/list.go:
```diff
@@ -159,16 +159,18 @@ func printResults(config listConfig, projects []queries.Project, owner string) e
+		config.tp.AddField(
+			strconv.Itoa(int(p.Number)),
+			tableprinter.WithTruncate(nil),
+			tableprinter.WithColor(cs.ColorFromString(format.ColorForProjectState(p))),
```
Adding the color to the number feels excessive since we are already coloring the state column. In this case I feel like having the color on the state column makes the most sense.

samcoe commented on pkg/cmdutil/json_flags.go:
```diff
@@ -110,6 +110,70 @@ func checkJSONFlags(cmd *cobra.Command) (*exportFormat, error) {
+			return cmd.Parent().FlagErrorFunc()(c, e)
+		}
+		return e
+	})
```
I don't think this is necessary.

samcoe commented on pkg/cmdutil/json_flags.go:
```diff
@@ -110,6 +110,70 @@ func checkJSONFlags(cmd *cobra.Command) (*exportFormat, error) {
 	return nil, nil
 }
 
+func AddFormatFlags(cmd *cobra.Command, exportTarget *Exporter, fields []string) {
```
What are your thoughts on just removing the `fields` argument since it is not actually being used? I think the `Exporter` interface should still work properly, it will just be up to the implementors of `ExportData` for each struct to understand that the `fields` argument will be nil. Seems like this would simplify using this function a good bit.

heaths replied:
I could try that. I looked through some `Exporter` implementors and wasn't sure off hand if I needed to support those in general.

samcoe commented on pkg/cmd/project/shared/format/json.go:
```diff
@@ -9,7 +9,7 @@ import (
 
 // JSONProject serializes a Project to JSON.
 func JSONProject(project queries.Project) ([]byte, error) {
```
All these `JSON...` functions I think can and should be refactored to utilize the `ExportData` pattern we have established else were in the codebase. For now just doing it for `Project` and `Projects` but doing it for other exported structs through the `project` commands.  

For example 
```go

//TODO: implement in queries.go near definition of Project
func (p Project) ExportData(_ []string) map[string]interface{} {
  return nil
}

//TODO: implement in queries.go near definition of Projects
func (p Projects) ExportData(_ []string) map[string]interface{} {
  return nil
}

func JSONProject(project []queries.Project) ([]byte, error) {
    return json.Marshal(project.ExportData(nil))
}

func JSONProject(projects []queries.Projects) ([]byte, error) {
    return json.Marshal(projects.ExportData(nil))
}

// In project list command
if config.opts.exporter != nil {
    // Would need to construct projects from []project and totalCount
    return config.opts.exporter.Write(config.IO, projects)
}

```

Basically I think all these structs should know how to serialize themselves into maps rather than having these helper function that do it. We can keep the helper functions in place for now but eventually I see them going away completely.

This does seem like a large refactor job though, so willing to hear differing opinions on this.

heaths replied:
That'd simplify things quite a bit. I was a little surprised to see a near-duplication of models, and this would clean it up quite a bit.

heaths replied:
I started on this but, you're right: it's a big refactoring. Perhaps in another PR just to isolate all the changes? I stashed away a good start on this for now.

samcoe reviewed (commented):
@heaths Thanks for iterating on this. I left a handful of comments that could change the implementation quite a bit so let me know your thoughts on them.

As for table output I personal really like having the table headers and I am guessing that there are people on both side of that preference. For now let's leave them in place without changing them and we can open another issue for the team and community to discuss what we TTY table output to look like. Don't want to bog down this PR with that discussion/work.

heaths commented:
> As for table output I personal really like having the table headers and I am guessing that there are people on both side of that preference. For now let's leave them in place without changing them and we can open another issue for the team and community to discuss what we TTY table output to look like. Don't want to bog down this PR with that discussion/work.

@samcoe I actually don't dislike the headers - in fact, I prefer them so the columns are more obvious - just was pointing out it's inconsistent. Maybe we should add them to all tables. I'll open another issue, though, on consistency and colorization of table headers.

samcoe commented:
@heaths Ah I see I misunderstood your previous comment, apologies for that. I would also like to see them added to all our table output to make it consistent 👍

heaths commented:
@samecoe no, I think you understood right. :smile:. I was pointing out they were inconsistent throughout the CLI, but I do think the headers make for a better UX. No guess work as to what columns are, even if most seem to be self-explanatory to me, anyway.

I did want to ask, though, only because I haven't dug deep into the `ExportData` functionality, are these two comments/suggestions in conflict?

1. https://github.com/cli/cli/pull/7832#discussion_r1338559707
2. https://github.com/cli/cli/pull/7832#discussion_r1338617300

Or not, since the latter would just fully de/serialize all its fields? I'm not entirely clear on how the passed-in `fields` parameter is used...rather, why it's passed down to other models' `ExportData`. When I get some time (hopefully this weekend), I'll dig in more, but thought I might ask here anyway in case a quick summary would help expedite my investigation and refactoring.

samcoe commented:
> Or not, since the latter would just fully de/serialize all its fields?

Exactly this. The structs that are exported from commands using the `format` flag instead of the `json` flag should implement the `exportable` interface but ignore the passed in `fields` argument as they will be serializing all the fields of the struct rather than relying on the passed in ones.

I am also open to keeping the list of fields, my intention of the comment was to simplify the implementation, but perhaps keeping consistency with other commands/implementations is more valuable here.

heaths commented:
> I am also open to keeping the list of fields, my intention of the comment was to simplify the implementation, but perhaps keeping consistency with other commands/implementations is more valuable here.

If it were a lib and you wanted a stable API, I'd agree. But you make a good point that anything using the `--format` flags (and friends) really doesn't need it. So with no filtering needed, I understand and appreciate your point now. Seems worth removing. Could always add it later through a fairly simple refactoring if needed.

samcoe commented:
@heaths I wanted to circle back around to this and see if there was anything blocking this from getting out of the draft state? I had put off continuing the review process as the PR state was set to work in process.

heaths commented:
No. I'll work on this and the other this weekend. I ended up not touching my computer much over the holiday, opting for some woodworking and other physical world projects instead.

heaths committed:
Remove unnecessary fields from JSON format flags

heaths committed:
Add standard format help to remaining commands

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

heaths commented:
@samcoe I completed what we discussed sans `ExportData`. As noted in the PR comments, you're right: this is a large refactoring. I'm happy to do it (perhaps after the JSON merge PR), but I suggest it be done in a separate PR to isolate the changes for easier review.

I was considering instead to just add `json` tags to the fields, but I see several models have flattened structures, like flattening the login for orgs and users. Though, I suppose I could extract that as a separate type and implement `MarshalJSON`. Or would you prefer `ExportData`? If we go that route, I'll probably add few helpers somewhere under `internal` to, for example, emulate `omitempty` for `json` tags.

heaths commented:
As I'm adding regression tests to make sure whatever course we take doesn't break existing output, I found a number of project commands aren't actually writing JSON for the return projects, but the minimal input projects (project ID, basically). I'm assuming this would be a desirable regression.

E.g., with `gh project close`:

```go
project, err := config.client.NewProject(canPrompt, owner, config.opts.number, false)
// --- CLIP ---
query, variables := closeArgs(config)

err = config.client.Mutate("CloseProjectV2", query, variables)
// ---- CLIP ---
if config.opts.exporter != nil {
	return printJSON(config, *project) // should be query.UpdateProjectV2.ProjectV2
}

return printResults(config, query.UpdateProjectV2.ProjectV2)
```

heaths committed:
Add JSON format regression tests

Also fixed a number of `project` commands that didn't format the right object as JSON.

heaths commented:
I'm going to pull in the changes from #8541 to fix the JSON output, which seems related more to this PR than refactoring how the data is serialized. Probably the regression tests as well, because I made sure to add more data to the responses to check most - though, not all - of the JSON model data; otherwise, what's the point of improving the JSON output if the `--format json`-formatted responses are useless for some `project` commands (about 30-40%).

samcoe commented on pkg/cmdutil/json_flags.go:
```diff
@@ -121,6 +173,18 @@ type exportFormat struct {
+}
+
+// JSONExporterOption customizes a JSON Exporter for NewJSONExporter.
+type JSONExporterOption func(e *exportFormat)
```
Looks like this is unused, none of the invocations of `NewJSONExporter` utilize any options. What was the occasion you anticipated using this?

heaths replied:
I figured they might coeme in handy later and since it's technically a public API, figured I'd get the signature right now (helping others avoiding breaking changes in Azure APIs is part of my day job). But, it's an EXE, so maybe this isn't important. I can remove it if you'd rather.

heaths replied:
Removed.

samcoe commented on pkg/cmdutil/json_flags.go:
```diff
@@ -121,6 +173,18 @@ type exportFormat struct {
 }
 
+// NewJSONExporter returns an Exporter to emit JSON.
+func NewJSONExporter(opts ...JSONExporterOption) Exporter {
```
As this appears to be only used in tests I am wondering if this helper could be moved to some place like `pkg/cmd/project/shared/format/test.go`?

heaths replied:
It needs to have access to the `exportFormat` struct, though - or should I just duplicate it? I was hoping to avoid the latter. Or do I make `exportFormat` public instead? That's really all that's needed - just to return an implementation of `cmdutil.Exporter`.

heaths replied:
I removed the options pattern, but kept the public function. And while I realize there's only one format supported now, I know more have been considered e.g., CSV. As such, I renamed the generic `exportFormat` to `jsonExporter`. I also considered just making the whole thing public, but perhaps a factory function would be better if ever needed - to switch on the format. Seemed overkill right now, though I'm happy to implement other format sometime.

Feel free to resolve if this is fine, or give me some other idea of what you'd like best.

samcoe commented on pkg/cmd/project/close/close_test.go:
```diff
@@ -453,3 +452,100 @@ func TestRunClose_Reopen(t *testing.T) {
+
+	err := runClose(config)
+	assert.NoError(t, err)
+	assert.JSONEq(
```
TIL `assert.JSONEq` 👍

heaths replied:
Yeah. Me too! I wanted something like this and was hoping not to have to write it myself, so typed `assert.` and scrolled through the methods.

samcoe commented on pkg/cmd/project/item-delete/item_delete.go:
```diff
@@ -126,6 +126,8 @@ func printResults(config deleteItemConfig) error {
 }
 
 func printJSON(config deleteItemConfig, ID githubv4.ID) error {
```
```suggestion
func printJSON(config deleteItemConfig, id githubv4.ID) error {
```

Small nit here, I know this was existing code, but since we are already here it makes sense to fix this.

samcoe commented on pkg/cmd/project/shared/format/display.go:
```diff
@@ -0,0 +1,19 @@
+	"github.com/cli/cli/v2/pkg/cmd/project/shared/queries"
+)
+
+func ProjectState(project queries.Project) string {
+	if project.Closed {
+		return "closed"
+	}
+	return "open"
+}
+
+func ColorForProjectState(project queries.Project) string {
+	if project.Closed {
+		return "gray"
+	}
+	return "green"
+}
```
Since we do not verify output coloring in our tests it would be nice to add some simple units tests for these functions.

samcoe commented on pkg/cmdutil/json_flags.go:
```diff
@@ -110,6 +110,58 @@ func checkJSONFlags(cmd *cobra.Command) (*exportFormat, error) {
+		}
+
+		if export, err := checkFormatFlags(c); err == nil {
+			if export == nil {
+				*exportTarget = nil
+			} else {
+				*exportTarget = export
+			}
```
```suggestion
			*exportTarget = export
```

I think this should work as expected.

heaths replied:
I thought that originally, but a variable that is nil is not the same as nil, it seems. I think it's because the return is `*jsonExporter` (nee `exportFormat`). It'd odd, I agree. The same goes for the [comment below](https://github.com/cli/cli/pull/7832#discussion_r1445521174) - it's related.

samcoe replied:
Ahh makes sense 👍 Good catch.

samcoe commented on pkg/cmdutil/json_flags_test.go:
```diff
@@ -116,6 +116,98 @@ func TestAddJSONFlags(t *testing.T) {
+				assert.EqualError(t, err, tt.wantsError)
+				return
+			}
+			if tt.wantsExport == nil {
+				assert.Nil(t, exporter)
+			} else {
+				assert.Equal(t, tt.wantsExport, exporter)
+			}
```
```suggestion
				assert.Equal(t, tt.wantsExport, exporter)
```

heaths replied:
`assert.Equal` checks strict equality, and `nil` !== `(*jsonExporter)nil`, it seems. See the [comment above](https://github.com/cli/cli/pull/7832#discussion_r1445821168) for more context.

heaths replied:
This explains the problem: https://go.dev/doc/faq#nil_error

It would seem to explain the other issue as well. Perhaps there's a better way of handling that in callers, but it works get more verbose throughout the codebase. Perhaps a comment with that link and brief explanation for future maintainers? The behavior is otherwise unintuitive.

samcoe reviewed (changes requested):
@heaths Wow, what a incredible amount of work on this! I added a couple inline comments that I would like to see addressed. Please let me know if you have any questions about them. I think this is very close to being shippable. I will take an initial look at https://github.com/cli/cli/pull/8541 tomorrow.

heaths committed:
Resolve PR feedback

samcoe reviewed (approved):
@heaths Thanks for making the requested changes. This looks ready to go! Really appreciate you taking the time to add in all those tests.
